### PR TITLE
Project Recovery: Remove saving of groups

### LINKED ISSUE #22922: Handle grouped workspaces in project recovery
Project recovery script fails if there is a group workspaces command in a workspace history, but one of the workspaces in the group has been deleted before the recovery files were written.

## Reproduce

* Before opening Mantid, set `projectRecovery.secondsBetween = 10` in `Mantid.properties`
* Run the following script
```
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws1')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws2')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws3') 
GroupWorkspaces(InputWorkspaces='ws,ws1,ws2,ws3', OutputWorkspace='NewGroup')
```
* Delete `ws3`
* Wait 10 seconds
* Run `Segfault`
* Restart Mantid - the recovery should fail

## Additional context

* This issue is likely to be valid for algorithms with `ADSValidator` in them. Speak to @NickDraper about this for more information.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
Changed multiple classes to stop project recovery saving grouped workspaces, this is useful because it stops a situation where project recovery does not work properly, when you delete a workspace or move it from outside of a group

**To test:**
* Before opening Mantid, set `projectRecovery.secondsBetween = 10` in `Mantid.properties`
* Run the following script
```
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws1')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws2')  
CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', Names='A,B,C', Units='U,U,U', OutputWorkspace='ws3') 
GroupWorkspaces(InputWorkspaces='ws,ws1,ws2,ws3', OutputWorkspace='NewGroup')
```
* Delete `ws3`
* Wait for the project to be saved
* Run `Segfault`
* Restart Mantid
* Click Yes on the project recovery pop up
* You should have all workspaces back outside of the original group

Fixes #22922 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/ScriptBuilder.cpp [hunk 1]
```diff
@@ -87,13 +88,23 @@ void ScriptBuilder::writeHistoryToStream(
     }
   } else {
     // create the string for this algorithm
-    os << buildAlgorithmString(*algHistory);
-    if (m_timestampCommands) {
-      os << " # " << algHistory->executionDate().toISO8601String();
+    if (!m_ignoreGroups) {
+      createStringForAlg(os, algHistory);
+    } else if (m_ignoreGroups && algHistory->name() != "GroupWorkspaces") {
+      createStringForAlg(os, algHistory);
```
Comment (soininen) [inline Framework/API/src/ScriptBuilder.cpp]: Could this `if`-`else` construct be simplified to:
```c++
if (!m_ignoreGroups || algHistory->name() != "GroupWorkspaces") {
  createStringForAlg(os, algHistory);
}
```
?

File: Framework/Algorithms/src/GeneratePythonScript.cpp [hunk 2]
```diff
@@ -44,9 +45,14 @@ void GeneratePythonScript::init() {
                   "Unroll all algorithms to show just their child algorithms.",
                   Direction::Input);
 
+  declareProperty("IgnoreGroups", false,
```
Comment (soininen) [inline Framework/Algorithms/src/GeneratePythonScript.cpp]: New properties should be declared at the end of `init()`. Adding properties in the middle will break Python scripts that don't use named arguments but rely on the order of the properties.

File: Framework/API/inc/MantidAPI/HistoryView.h [hunk 3]
```diff
@@ -65,6 +65,12 @@ class MANTID_API_DLL HistoryView {
   const std::vector<HistoryItem> &getAlgorithmsList() const {
     return m_historyItems;
   };
+
+  std::shared_ptr<std::vector<Mantid::API::HistoryItem>>
+  getMutableAlgorithmsList() {
```
Comment (soininen) [inline Framework/API/inc/MantidAPI/HistoryView.h]: Is this function used anywhere?

**Review by soininen — CHANGES_REQUESTED** [review]
I followed the test procedure in the description and did some fiddling with the workspace group on my own. Things seemed to work fine :+1: There are some issues with the code changes, see below, though generally the changes look good. What concerns me is the lack of unit tests. Was there some reason none were added/modified in this PR?

**Comment by Pasarus** [comment]
It is hard to test in project recovery because of how it works, thus tests are usually handled manually. That is why none were added.

File: Framework/API/inc/MantidAPI/HistoryView.h (hunk 3 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/HistoryView.h]: Nope, it was left over from a previous fix that was later re-wrote. Removed.

File: Framework/API/src/ScriptBuilder.cpp (hunk 1 above)
Comment (Pasarus) [reply Framework/API/src/ScriptBuilder.cpp]: It can, good spot!

File: Framework/Algorithms/src/GeneratePythonScript.cpp (hunk 2 above)
Comment (Pasarus) [reply Framework/Algorithms/src/GeneratePythonScript.cpp]: Resolved this now.

**Review by soininen — APPROVED** [review]

File: MantidPlot/src/ProjectRecovery.cpp [hunk 4]
```diff
@@ -490,7 +492,7 @@ void ProjectRecovery::saveWsHistories(const Poco::Path &historyDestFolder) {
   const auto &ads = Mantid::API::AnalysisDataService::Instance();
 
   // Hold a copy to the shared pointers so they do not get deleted under us
-  const auto wsHandles =
+  std::vector<boost::shared_ptr<Mantid::API::Workspace>> wsHandles =
```
Comment (martyngigg) [inline MantidPlot/src/ProjectRecovery.cpp]: `const auto` was following the recommended practice post C++11. Is there some reason I've missed that this no longer works?

**Review by martyngigg — COMMENTED** [review]
Is this slightly brutal? Even if your `WorkspaceGroup` was going to be brought back in tact, it now won't be?

**Comment by Pasarus** [comment]
So I spoke with Kieth quite a lot about a better way to sort it, and at that time and at present we don't have the time to find a better fix. Right now project recovery straight up fails in this case and stops it from autosaving in that instance of mantid, which is a critical failure, this was a relatively quick fix that would stop it from being completely broken in the case of a workspace group. We do intend to make this better in the future.

**Comment by martyngigg** [comment]
:+1: I hadn't appreciated it kills the saving completely.

**Pull request merged**